File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedLevelEvent.java
package net.easecation.ghosty.recording.level.updated;

import cn.nukkit.level.Level;
import cn.nukkit.math.Vector3f;
import cn.nukkit.network.protocol.LevelEventPacket;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.recording.level.LevelRecordNode;
import org.itxtech.synapseapi.multiprotocol.protocol112.protocol.LevelEventPacket112;
import org.itxtech.synapseapi.multiprotocol.protocol116100.protocol.LevelEventPacket116100;
import org.itxtech.synapseapi.multiprotocol.protocol14.protocol.LevelEventPacket14;
import org.itxtech.synapseapi.multiprotocol.protocol16.protocol.LevelEventPacket16;
import org.itxtech.synapseapi.multiprotocol.protocol17.protocol.LevelEventPacket17;

public class LevelUpdatedLevelEvent implements LevelUpdated {

    private int evid;
    private float x = 0;
    private float y = 0;
    private float z = 0;
    private int data = 0;

    private LevelUpdatedLevelEvent(int evid, float x, float y, float z, int data) {
        this.evid = evid;
        this.x = x;
        this.y = y;
        this.z = z;
        this.data = data;
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket14 packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket16 packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket17 packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket112 packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public static LevelUpdatedLevelEvent of(LevelEventPacket116100 packet) {
        return new LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data);
    }

    public LevelUpdatedLevelEvent(BinaryStream stream) {
        this.read(stream);
    }

    @Override
    public int getUpdateTypeId() {
        return LevelUpdated.TYPE_LEVEL_EVENT;
    }

    @Override
    public void processTo(LevelRecordNode node) {
        LevelEventPacket packet = new LevelEventPacket();
        packet.evid = this.evid;
        packet.x = this.x;
        packet.y = this.y;
        packet.z = this.z;
        packet.data = this.data;
        node.handleLevelChunkPacket(Level.chunkHash((int) x >> 4, (int) z >> 4), packet);
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putVarInt(evid);
        stream.putVector3f(x, y, z);
        stream.putVarInt(data);
    }

    @Override
    public void read(BinaryStream stream) {
        this.evid = stream.getVarInt();
        Vector3f vector3f = stream.getVector3f();
        this.x = vector3f.x;
        this.y = vector3f.y;
        this.z = vector3f.z;
        this.data = stream.getVarInt();
    }

    @Override
    public String toString() {
        return "LevelUpdatedLevelEvent{" +
            "evid=" + evid +
            ", x=" + x +
            ", y=" + y +
            ", z=" + z +
            ", data=" + data +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedLevelEvent.kt
package net.easecation.ghosty.recording.level.updated

import cn.nukkit.level.Level
import cn.nukkit.math.Vector3f
import cn.nukkit.network.protocol.LevelEventPacket
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.recording.level.LevelRecordNode
import org.itxtech.synapseapi.multiprotocol.protocol112.protocol.LevelEventPacket112
import org.itxtech.synapseapi.multiprotocol.protocol116100.protocol.LevelEventPacket116100
import org.itxtech.synapseapi.multiprotocol.protocol14.protocol.LevelEventPacket14
import org.itxtech.synapseapi.multiprotocol.protocol16.protocol.LevelEventPacket16
import org.itxtech.synapseapi.multiprotocol.protocol17.protocol.LevelEventPacket17

@Serializable
data class LevelUpdatedLevelEvent(
    val evid: Int,
    val x: Float,
    val y: Float,
    val z: Float,
    val data: Int,
) : LevelUpdated {

    override fun getUpdateTypeId(): Int = LevelUpdated.TYPE_LEVEL_EVENT

    override fun processTo(node: LevelRecordNode) {
        LevelEventPacket().also {
            it.evid = this.evid
            it.x = this.x
            it.y = this.y
            it.z = this.z
            it.data = this.data
        }.also { packet ->
            node.handleLevelChunkPacket(Level.chunkHash((x.toInt() shr 4), (z.toInt() shr 4)), packet)
        }
    }

    override fun backwardTo(node: LevelRecordNode) = Unit

    companion object {
        @JvmStatic
        fun of(packet: LevelEventPacket) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        fun of(packet: LevelEventPacket14) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        fun of(packet: LevelEventPacket16) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        fun of(packet: LevelEventPacket17) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        fun of(packet: LevelEventPacket112) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        fun of(packet: LevelEventPacket116100) =
            LevelUpdatedLevelEvent(packet.evid, packet.x, packet.y, packet.z, packet.data)

        @JvmStatic
        val ADAPTER: LevelUpdateAdapter<LevelUpdatedLevelEvent> = Adapter
    }

    private object Adapter : LevelUpdateAdapter<LevelUpdatedLevelEvent> {
        override fun write(updated: LevelUpdatedLevelEvent, stream: BinaryStream) {
            stream.putVarInt(updated.evid)
            stream.putVector3f(updated.x, updated.y, updated.z)
            stream.putVarInt(updated.data)
        }

        override fun read(stream: BinaryStream, formatVersion: Int): LevelUpdatedLevelEvent {
            val evid = stream.varInt
            val vector3f = stream.vector3f
            val data = stream.varInt
            return LevelUpdatedLevelEvent(evid, vector3f.x, vector3f.y, vector3f.z, data)
        }
    }
}